code-search: improve actions discoverability and UX on file page

BolajiOlajide opened the pull request:
Closes #57702

This PR revamps the actions on the code search page.

## Before

![CleanShot 2023-11-24 at 22 48 35@2x](https://github.com/sourcegraph/sourcegraph/assets/25608335/a5e373c7-8be2-4e37-aae1-fd3cab1c5c82)

## After

https://github.com/sourcegraph/sourcegraph/assets/25608335/0a3e0a15-fe5c-4445-b61f-0698d1d3a497


## Test plan

* Navigate to any file in a repository. (Example: [<SOURCEGRAPH_URL>/github.com/sourcegraph/sourcegraph@main/-/blob/.graphqlrc.yml](https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph@main/-/blob/.graphqlrc.yml))

* The actions navigation should display a label.


cc @rrhyne for QA.

sqs commented:
Woo, this is an awesome improvement! Thank you!

sqs commented:
A few issues I see:

On narrow windows, the items overlap the file path:

![image](https://github.com/sourcegraph/sourcegraph/assets/1976/75b90df5-a647-4e5d-b2fc-1b4b223fc1bb)

Generally, the buttons take up a lot of width. Not sure of the solution here.

The "grammar" of the buttons is not consistent. `Ask Cody` is a verb phrase, but `Editor` is a noun, for example.

sqs commented on client/web/src/repo/blob/actions/ToggleHistoryPanel.module.scss:
```diff
@@ -0,0 +1,17 @@
+@import 'wildcard/src/global-styles/breakpoints';
+
+.repo-action-label {
```
This same CSS needs to apply to all of the icon actions, so it and the corresponding TSX code should be factored out into a separate component.

rrhyne commented:
Thanks @sqs we discussed this issue before break and landed on a few solutions:

* At smaller breakpoints do not show labels ([figma](https://www.figma.com/file/skTkU7PPWCvH2SAlP9GYtC/Search-Suite-Q4-UX-work?type=design&node-id=374-9034&mode=design&t=e6i07IW9pVXyZZEO-4))
  * This is counter one of the problems we are solving, but we need to compensate for small screens in some way
* On really small screens, drop icons to their own line ([figma](https://www.figma.com/file/skTkU7PPWCvH2SAlP9GYtC/Search-Suite-Q4-UX-work?type=design&node-id=842-23151&mode=design&t=e6i07IW9pVXyZZEO-4))

BolajiOlajide commented:
I fixed the overlap in my recent commit.

![CleanShot 2023-11-28 at 03 19 48](https://github.com/sourcegraph/sourcegraph/assets/25608335/12f74e47-6fbb-4024-9a09-fa6b930be52b)

I also fixed the icon arrangement on mobile view:

![CleanShot 2023-11-28 at 03 27 10@2x](https://github.com/sourcegraph/sourcegraph/assets/25608335/a11126bd-caac-4c89-af06-7179bcce0282)

BolajiOlajide commented on client/web/src/repo/RepoActionInfo.tsx:
```diff
@@ -0,0 +1,56 @@
+}) => {
+    let iconToBeRendered: JSX.Element | undefined | null
+
+    if (typeof icon === 'string') {
+        iconToBeRendered = (
+            <Icon svgPath={icon} aria-hidden={true} className={classNames(styles.repoActionIcon, iconClassName)} />
+        )
+    } else if (React.isValidElement(icon) || icon === null) {
+        iconToBeRendered = icon
+    } else {
+        // we cast the icon type here because there's no valid way to check that `icon` is explicitly of type
+        // MdiReactIconComponentType | ComponentType<{ className?: string | undefined }>
+        // especially considering `React.ReactNode` is a union of all primitive types and normal react
+        // components.
+        // This is a hack and should be reworked later.
+        iconToBeRendered = (
+            <Icon
+                as={icon as ComponentType}
+                aria-hidden={true}
+                className={classNames(styles.repoActionIcon, iconClassName)}
+            />
+        )
+    }
```
I'm trying to avoid the style duplication but tbh it's beginning to look like the cost isn't worth it .

BolajiOlajide commented on client/web/src/repo/RepoHeaderContextMenu.tsx:
```diff
@@ -0,0 +1,19 @@
+export const RepoHeaderContextMenu: FC<RepoHeaderContextMenuProps> = ({ actions }) => (
+    <Menu>
+        <MenuButton variant="secondary" outline={true} className="py-0">
+            ...
```
That make sense. I'll update.

BolajiOlajide commented on client/web/src/repo/actions/CopyPermalinkAction.tsx:
Nothing special happens. I just use that to know when the page might have a permalink and when it already is.

1. 	`sourcegraph.com/github.com/sourcegraph/sourcegraph@2f6286d5f69fc6168ace2ec5d784872815c2223b/-/blob/client/web/src/auth.ts`

2. `/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/auth.ts`

BolajiOlajide commented on client/web/src/repo/blob/GoToRawAction.module.scss:
Not very easy because some of the components in the ContextMenu have different styles.

vovakulikov reviewed (approved):
Tech changes are okay, overall the codebase there is very complex and I think we could do something about it but I see that at the moment it probably would take more time than we have. 

From layout perspective I think I got the state where wrapping doesn't work properly 
<img width="1358" alt="Screenshot 2023-12-21 at 11 26 21" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/c2545eda-2a0d-4b14-bfdb-8e837e005cfb">

And I think there is one problem with "Ask Cody" button in mobile layout in three dot context menu, this is something that Quinn reported above.

BolajiOlajide commented:
> Tech changes are okay, overall the codebase there is very complex and I think we could do something about it but I see that at the moment it probably would take more time than we have. 
> 
> 
> 
> From layout perspective I think I got the state where wrapping doesn't work properly 
> 
> <img width="1358" alt="Screenshot 2023-12-21 at 11 26 21" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/c2545eda-2a0d-4b14-bfdb-8e837e005cfb">
> 
> 
> 
> And I think there is one problem with "Ask Cody" button in mobile layout in three dot context menu, this is something that Quinn reported above.
> 
> 

That's interesting. I hadn't tested the wrapping while the AskCody sidebar is open. Because we currently hide the labels at about 900px and below, but the ask cody sidebar takes quite a lot of space, we might need to lower that a bit or use another means to calculate the width.
